### DISCUSSION TITLE
SoftSPI and sdcard on M5Stack Core2

### DISCUSSION THREAD

**Discussion by rgroener** [micropython/micropython#9184] [Libraries & Drivers] [answered]
Hi
I have an issue with initialising an sdcard on my M5Stack Core 2.
Since the pins connected to the sdcard do not correspond to one of the SPI slots, I tried to use the SoftSpi.
I tried to use the pins from the official M5Stack Core 2 Documentation, but had no luck. [(https://docs.m5stack.com/en/core/core2)]
According to other threads, on the last M5Stack, the pins corresponded to a SPI slot, therefore using the SoftSPI was not necessary...

sdcard is formatted with FAT32.
I tried different sdcards (2GB, 8GB, 16GB), no difference...

Does anyone see an obvious problem with my code?
Thanks in advance.
```py
from machine import SoftSPI,Pin
import sdcard
import uos
'''
micdroSDCard Pins on M5Stack
mosi = GPIO23
miso = GPIO38
sck  = GPIO18
cs   = GPIO4
'''
cs = Pin(4)
spi =SoftSPI(baudrate=100000,
             polarity=1,
             phase=0,
             sck=Pin(18),
             mosi=Pin(23),
             miso=Pin(38))

sd = sdcard.SDCard(spi,cs)
uos.mount(sd,"/sd")
uos.listdir('/sd')

file = open("/sd/data.txt", "w")
file.write("Test Test Test")
file.close()
```

**Comment by jimmo** [comment]
This is similar to https://forum.micropython.org/viewtopic.php?f=15&t=12988 ?

> do not correspond to one of the SPI slots, I tried to use the SoftSpi.

I'm not quite sure I follow... the ESP32 allows you to use any pins with a given hardware SPI.

**Comment by rgroener** [comment]
Yes, it is the same. I was unsure where to post...
Is it better to go on here or on the micropython forum?

> I'm not quite sure I follow... the ESP32 allows you to use any pins with a given hardware SPI.

Thanks for this information. It was not clear to me when reading the documentation. I am new to the ESP32, coming from the AVR processor Family.... It is a bit confusing that you can change pins on a hardware SPI. Must have skipped this information:-(

I imported the newest sdcard.py file, but an older sdcard.py file compiled into the firmware was used when calling sdcard functions. The firmware version seems to be outdated too...

I will try it with the newest sdcard file and hardware SPI.

**Reply by jimmo** [reply]
> Is it better to go on here or on the micropython forum?

here. we're in the process of moving the old forum to here.

> I will try it with the newest sdcard file and hardware SPI.

I missed something in your first post... on ESP32 you should use the built-in SDCard class on the `machine` module. Do not use the `sdcard.py`.

**Comment by rgroener** [comment]
Isn't the old version a problem when using the machine module? 

> MicroPython v1.16-115-g6a9133a8b-dirty on 2021-07-20; M5STACK Core2 with ESP32

I have never compiled a firmware for an esp32...

Currently, with the newest sdcard.py file, the output is: 

> OSError: timeout waiting for v2 card

I will try it with the machine module.

**Comment by rgroener** [comment]
Version with the machine module. 
```py
'''
micdroSDCard Pins on M5Stack
mosi = GPIO23
miso = GPIO38
sck  = GPIO18
cs   = GPIO4
'''
import machine, os

sd = machine.SDCard(slot=2,sck=machine.Pin(18),miso=machine.Pin(38),mosi=machine.Pin(23),cs=machine.Pin(4))
os.mount(sd, '/sd')  # mount

os.listdir('/sd') 
file = open("/sd/data.txt", 'w')
file.write("Test Test Test")
file.close()
os.umount('/sd') 
```

First run after inserting card and hw reset gives:

```
>   File "<stdin>", line 11, in <module>
OSError: 16
```

Continuous runs give:

```
>   File "<stdin>", line 10, in <module>
OSError: (-259, 'ESP_ERR_INVALID_STATE')
```

(Edit by @jimmo to add code formatting)

**Reply by jimmo** [reply]
`OSError: 16` means that we're failing to call `sdmmc_card_init`.

It might be worth trying to turn the frequency down a bit. Add `freq=1_000_000` to the `machine.SDCard(...)`.

FYI please see instructions on formatting code snippets in https://github.com/orgs/micropython/discussions/9111

**Comment by rgroener** [comment]
> FYI please see instructions on formatting code snippets in https://github.com/orgs/micropython/discussions/9111

Thanks for the hint, I will do so...

Lowering the frequency did not change it, still `OSError: 16`

```py
sd = machine.SDCard(slot=2,freq=1_000_000, sck=machine.Pin(18),miso=machine.Pin(38),mosi=machine.Pin(23),cs=machine.Pin(4))
```

**Reply by jimmo** [reply]
hrmm, I'm starting to wonder if maybe there's something wrong with the ESP32 built-in SDCard driver (well, the MicroPython bindings for it). See also this thread here: https://forum.micropython.org/viewtopic.php?f=18&t=12735&

Unfortunately I don't think I have an ESP32 board with an SDCard reader on it. I will investigate.

In theory you should be able to use sdcard.py as a drop-in replacement for the built-in one. But it's curious that doesn't work for you either. I think I'm going to have to track down a board.

**Comment by jimmo** [comment, marked as answer]
@rgroener I managed to track down an SD adaptor.

I have tested sdcard.py on an ESP32 using hardware SPI(1), SPI(2) and SoftSPI.

```
MPY: soft reboot
MicroPython v1.19.1-278-g6dfb462ef on 2022-08-13; TinyPICO with ESP32-PICO-D4
Type "help()" for more information.
>>> import machine, sdcard, os
>>> spi = machine.SPI(2, sck=machine.Pin(5), mosi=machine.Pin(22), miso=machine.Pin(21))
>>> sd = sdcard.SDCard(spi, machine.Pin(18))
>>> os.mount(sd, '/sd')
>>> os.listdir('/sd')
['demo.txt']
>>> os.umount('/sd')
```

Changing to

```
spi = machine.SoftSPI(sck=machine.Pin(5), mosi=machine.Pin(22), miso=machine.Pin(21))
```

works fine for software SPI.

I have also tested with `machine.SDCard`, and that seems fine too.

```
MPY: soft reboot
MicroPython v1.19.1-278-g6dfb462ef on 2022-08-13; TinyPICO with ESP32-PICO-D4
Type "help()" for more information.
>>> import machine, os
>>> sd = machine.SDCard(slot=2, sck=machine.Pin(5), mosi=machine.Pin(22), miso=machine.Pin(21), cs=machine.Pin(18))
>>> os.mount(sd, '/sd')
>>> os.listdir('/sd')
['demo.txt']
>>> os.umount('/sd')
>>> 
```

Unfortunately I don't have an ESP32 board that exposes pin 38 so can't try your exact configuration.

So... not sure. Does the M5 Stack have some additional requirement like turning on power to the SDCard? (Many boards have something like this).

**Reply by jimmo** [reply]
@rgroener 

> Does the M5 Stack have some additional requirement like turning on power to the SDCard?

Answering my own question... the answer is yes.

https://m5stack.oss-cn-shenzhen.aliyuncs.com/resource/docs/schematic/Core/CORE2_V1.0_SCH.pdf

You need to enable PERI_VDD, which powers the SD Card.

This is done by talking to the power management chip over I2C. You can see the M5 Arduino code here:
https://github.com/m5stack/M5Core2/blob/master/src/AXP192.cpp

Here's an (untested) attempt to port that to MicroPython (warning this code messages with the power management chip and is untested!).

```py
i2c_axp192 = machine.I2C(0, sda=machine.Pin(21), scl=machine.Pin(22))

def read8bit(reg):
    print("read", reg)
    return i2c_axp192.readfrom_mem(0x34, reg, 1)[0]

def write1Byte(reg, v):
    print("write", reg, v)
    i2c_axp192.writeto_mem(0x34, reg, bytes((v,)))

def calcVoltageData(value, maxv, minv, step):
    data = 0;
    if value > maxv:
      value = maxv
    if value > minv:
      data = (value - minv) / step
    return int(data)

def setLDOVoltage(number, voltage):
    vdata = calcVoltageData(voltage, 3300, 1800, 100) & 0x0F;
    if number == 2:
        write1Byte(0x28, (read8bit(0x28) & 0x0F) | (vdata << 4));
    elif number == 3:
        write1Byte(0x28, (read8bit(0x28) & 0xF0) | vdata);

def setPeripherialsPower(state):
  if state:
    write1Byte(0x10, read8bit(0x10) | 0X02);
  else:
    write1Byte(0x10, read8bit(0x10) & 0XFB);

write1Byte(0x30, (read8bit(0x30) & 0x04) | 0x02)
setLDOVoltage(2, 3300)
setPeripherialsPower(True)
```

**Reply by jimmo** [reply]
Also worth looking at https://github.com/lukasmaximus89/M5Stick-C-Micropython-1.12/blob/master/axp192.py (although it's for a different board).

**Reply by rgroener** [reply]
jimmo, thanks a lot for your effort!

PERI_VDD is also used for the  display and is now enabled =>  display works...
Display and sdcard are on the same SPI (different cs)...
I also tried the SoftSPI, and different sdcards... => same output on REPL.
Due to the old micropython version in the firmware, the  sdcard.py is used 
[https://github.com/micropython/micropython-lib/blob/58f8bec54d5b3b959247b73a6e8f28e8493bd30b/micropython/drivers/storage/sdcard/sdcard.py](url)

```py
from machine import Pin, SPI
import machine
import time
import os
import axp202c
import sdcard
import ili9342c
import chango_32 as font_32

# Turn display power / enable PERI_VDD
axp = axp202c.PMU(address=0x34)
axp.enablePower(axp202c.AXP192_LDO2)

# Set backlight voltage
axp.setDC3Voltage(2800)
#initialize spi bus
spi = SPI(2,baudrate=1_000_000,sck=Pin(18),mosi=Pin(23),miso=Pin(38))
# initialize display
tft = ili9342c.ILI9342C(spi,320,240,reset=Pin(33, Pin.OUT),
    cs=Pin(5, Pin.OUT),
    dc=Pin(15, Pin.OUT),
    rotation=0)
      
tft.init()
tft.fill(ili9342c.BLACK)
time.sleep(1)
tft.rotation(0)
tft.write(font_32, "Display-Text", 20, 150, ili9342c.WHITE)      # and write the string

#spi = SPI(2, sck=Pin(18), mosi= Pin(23), miso= Pin(38))
sd = sdcard.SDCard(spi, cs= Pin(4))
os.mount(sd, '/sd')
os.listdir('/sd')
file = open("/sd/data.txt", 'w')
file.write("Test Test Test")
file.close()
os.umount('/sd')
```
output is now:
```
MicroPython v1.16-115-g6a9133a8b-dirty on 2021-07-20; M5STACK Core2 with ESP32
Type "help()" for more information.
>>> %Run -c $EDITOR_CONTENT
* initializing pins
* initializing i2c
* initializing mpu
* Detect PMU Type is AXP192
Traceback (most recent call last):
  File "<stdin>", line 32, in <module>
  File "sdcard.py", line 240, in readblocks
  File "sdcard.py", line 181, in readinto
OSError: timeout waiting for response
>>> 
```

**Reply by jimmo** [reply]
> Due to the old micropython version in the firmware, the sdcard.py is used

Not sure what you mean by this. It's not part of the firmware. Just grab the latest version and copy it to the filesystem.

**Reply by rgroener** [reply]
Sorry for being unclear and my ignorance...
I used a pre-compiled firmware [https://github.com/russhughes/ili9342c_mpy/tree/main/firmware/M5CORE2](url)

> ... Just grab the latest version and copy it to the filesystem.

I assume, this only works if I compile the firmware myself?

I downloaded the latest micropython version from github and tried to compile it.
It worked... At least REPL  shows:

> MicroPython v1.19.1-389-g4903e48e3 on 2022-09-09; ESP32 module with ESP32

With this version, reading and writing to the sdcard works.
```py
from machine import SPI
import axp202c
import machine, os

axp = axp202c.PMU(address=0x34)
#enable PERI_VDD (powers display and sdcard)
axp.enablePower(axp202c.AXP192_LDO2)

sd = machine.SDCard(slot=2, sck=machine.Pin(18), mosi=machine.Pin(23), miso=machine.Pin(38), cs=machine.Pin(4))
os.mount(sd, '/sd')
print(os.listdir('/sd'))
file = open("/sd/data.txt", 'a')
file.write("Bla Bla Bla")
file.close()
```
REPL
```
MicroPython v1.19.1-389-g4903e48e3 on 2022-09-09; ESP32 module with ESP32
Type "help()" for more information.
>>> %Run -c $EDITOR_CONTENT
* initializing pins
* initializing i2c
* initializing mpu
* Detect PMU Type is AXP192
['sd', 'data.txt']
>>> 
```
Now I have to find out how to compile the newest micropython version including the ili9342c driver or find the driver as module.
I will dig into it...

Thanks again for your big help!
Should I mark the thread as solved?

**Reply by jimmo** [reply]
OK I had no idea you weren't running an official build of MicroPython...

Glad it's resolved!
